### ISSUE TITLE
Improve FHIR Search Performance for Multiple Params and Values

### ISSUE THREAD

**Issue by alexanderkiel** [samply/blaze#799]
**TL;DR**

We will use streaming intersection and union on token search parameters to implement multiple search parameters and multiple values.

# Speedup

All queries are executed on the `1M` dataset and system `A5N46`. More information under [FHIR Search Performance](https://samply.github.io/blaze/performance/fhir-search.html).

**Query 1:**
 
```
[base]/Observation?status=final&category=vital-signs&code=<5 LOINC codes>
```

| Execution | Old time (s) | New time (s) | Speedup |
|-----------|------------:|--------------:|---------:|
| count        |              170 |                 2.61 |           65 |
| download |            1349 |                52.6 |           25 |         

**Query 2:**

```
[base]/Observation?code=<20 LOINC codes>&patient=<1000 patient IDs>
```

| Execution | Old time (s) | New time (s) | Speedup |
|-----------|------------:|--------------:|---------:|
| count       |                    1 |                      1 |              1 |
| download |              552 |                 52.6 |           10 |         

**Query 3:**

```
[base]/Observation?category=laboratory&date=<year>
```

| Execution | Old time (s) | New time (s) | Speedup |
|-----------|------------:|--------------:|---------:|
| count       |                117 |                  101 |         1.16 |

# Common Considerations

Currently the processing of the first search parameter results in a reducible collection of resource handles. However, in order to obtain a resource handle, a seek into the ResourceAsOf index is needed. In case a resource handle will be filtered out by later search parameters, this seek was wasteful. In this work, we will switch to use a data structure called **index handles** while processing search parameters and only obtain the resource handles in the very last moment after all matches are done. A index handle consists of the resource id and a list of hash prefixes and can be created by just scanning the SearchParamValueResource index.

# Improvement on multiple search parameters

Currently we scan through the SearchParamValueResource index for the first search parameter and seek into the ResourceSearchParamValue index for the rest of the search parameters. This approach only performs well if the first search parameter discriminates very well so that the number of required seeks is low. Choosing the right first search parameter is critical but we don't have statistics for that. Also in case multiple search parameters discriminate about the same, too many seeks are required.

However, at least for token search parameters we have a much better option. Token search parameters only allow point lookup, so the index handles of a single value are always ordered. That means, a scan of the SearchParamValueResource index for a single value will result in a reducible collection of ordered index handles. If we have more than one token search parameter, we can simply calculate the intersection of the reducible collections in a streaming fashion. Unfortunately we can't use transducers for that, because transducers only work on one stream. Instead we have to use iterators.

All other search parameter types allow for more than point lookups. In case more than one value is covered by a scan, our index handles are no longer ordered. In case we have only a few values, we could calculate the union of the streams.  But we might not do that in this task.

# Improvement on multiple values of one search parameter

In case more than one value is requested for a token search parameter via the comma syntax like `code=a,b` we can't simply concatenate the scans of the SearchParamValueResource index because one resource can match both values which would result in duplicate results. The current implementation just uses the `distinct` transducer to remove duplicates. However that means that we can't use the `start-id` to seek into the next page, because we no longer have the head of the steam as base to remove later duplicates. So the current implementation has to start always at the very beginning for all pages. That means that later pages will become very slow.

However, because the scans of individual values are ordered, we can implement a streaming union of the streams of multiple values. Also we can start each stream with the `start-id` because of the ordering.
